app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+                )
+            }
+        }
+    }
+
+    private fun setAndCalcTokenAmount(targetTokenType: StrigaOffRampTokenType, amount: BigDecimal) {
+        val calculatedAmount = interactor.calculateAmountByRate(targetTokenType, rate, amount)
+        setTokenAmount(targetTokenType, calculatedAmount)
+    }
+
+    private fun validate() {
+        val buttonState = interactor.validateAmount(
+            amountA = inputAmountA,
+            amountB = inputAmountB,
+            balance = balance
+        )
+        setButtonState(buttonState)
+
+        if (buttonState.isAmountError) {
+            view?.setTokenAErrorState(isError = true)
+        } else {
+            view?.setTokenAErrorState(isError = false)
+        }
+    }
+
+    private fun setButtonState(state: StrigaOffRampButtonState) {
+        timber.d("Set button state: $state")
+        viewState.value = viewState.value.copy(buttonState = state)
+    }
+
+    private fun setExchangeRateState(state: SwapRateTickerState) {
+        viewState.value = viewState.value.copy(exchangeRateState = state)
+    }
+
+    @NoCoverage
+    private fun logError(throwable: Throwable) {
+        when (throwable) {
+            is CancellationException -> {
+                if (!connectionManager.connectionStatus.value) {
+                    timber.i(throwable, "Canceled rate polling due to no internet connection")
+                } else {
+                    timber.i(
+                        throwable,
+                        "Canceled rate polling due to coroutine has been disposed"
+                    )
+                }
+            }
+            is StrigaDataLayerError.InternalError -> {
+                if (throwable.cause != null) {
+                    logError(throwable.cause!!)
+                } else {
+                    timber.e(throwable, "DataLayer: Error while rate polling")
+                }
+            }
+            else -> {
+                timber.e(throwable, "Error while rate polling")
+            }
+        }
+    }
+
+    private fun handleTokenState(newState: UserTokensState) {
+        when (newState) {
+            is UserTokensState.Idle -> Unit
+            is UserTokensState.Loading -> {
+                timber.d("Loading USDC balance")
+                balance = BigDecimal.ZERO
+                setTokenABalance(null)
+            }
+            is UserTokensState.Refreshing -> Unit
+            is UserTokensState.Error -> view?.showErrorMessage(newState.cause)
+            is UserTokensState.Empty -> {
+                timber.d("USDC balance is empty...")
+                balance = BigDecimal.ZERO
+                setTokenABalance(balance)
+            }
+            is UserTokensState.Loaded -> {
+                val usdc = newState.solTokens.find { it.isUSDC }
+                val balance = usdc?.total ?: BigDecimal.ZERO
+                timber.d("USDC balance = ${balance.formatFiat()}")
+                this.balance = balance
```

Comment:
this.usdcBalance

can be refactored to other token if we are going to support not USDC, but nah, i don't think so